tla-io/src/main/scala/at/forsyte/apalache/io/lir/ItfCounterexampleWriter.scala

```diff
@@ -0,0 +1,137 @@
+
+  /**
+   * Produce a JSON representation of a counterexample in the ITF format
+   *
+   * @param rootModule the module that produced the counterexample
+   * @param states     a sequence of next states
+   * @return the JSON representation of the counterexample in the ITF format
+   */
+  def mkJson(rootModule: TlaModule, states: List[NextState]): ujson.Value = {
+    // merge constant initialization and variable initialization into a single state
+    val state0 = states match {
+      case Seq(constInit)                => constInit._2
+      case Seq(constInit, initState, _*) => constInit._2 ++ initState._2
+      case Seq()                         => throw new IllegalArgumentException("Expected at least one state, found none")
+    }
+    val mappedStates = state0 :: states.drop(2).map(_._2)
+    // construct the root JSON object
+    val rootMap: mutable.LinkedHashMap[String, ujson.Value] = mutable.LinkedHashMap()
+    rootMap.put("#meta",
+        ujson.Obj(
+            "format" -> "ITF",
+            "format-description" -> "https://apalache.informal.systems/docs/adr/015adr-trace.html",
+            "description" -> "Created by Apalache on %s".format(Calendar.getInstance().getTime)
+        ))
+    paramsToJson(rootModule).foreach(params => rootMap.put("params", params))
+    rootMap.put("vars", varsToJson(rootModule))
+    rootMap.put("states", ujson.Arr(mappedStates.zipWithIndex.map((stateToJson _).tupled): _*))
+    ujson.Obj(rootMap)
+  }
+
+  override def write(rootModule: TlaModule, notInvariant: NotInvariant, states: List[NextState]): Unit = {
+    writer.write(ujson.write(mkJson(rootModule, states), indent = 2))
+  }
+
+  private def varsToJson(root: TlaModule): ujson.Value = {
+    val names = root.declarations.collect { case TlaVarDecl(name) =>
+      ujson.Str(name)
+    }
+    ujson.Arr(names: _*)
+  }
+
+  private def paramsToJson(root: TlaModule): Option[ujson.Value] = {
+    val names = root.declarations.collect { case TlaConstDecl(name) =>
+      ujson.Str(name)
+    }
+    if (names.isEmpty) None else Some(ujson.Arr(names: _*))
+  }
+
+  private def stateToJson(state: Map[String, TlaEx], index: Int): ujson.Value = {
+    val meta = ujson.Obj("index" -> ujson.Num(index))
+    val map = state.toList.sortBy(_._1).map(p => (p._1, exToJson(p._2)))
+    ujson.Obj("#meta" -> meta, map: _*)
+  }
+
+  private def exToJson: TlaEx => ujson.Value = {
+    case ValEx(TlaInt(num)) =>
+      if (num >= MIN_JS_INT && num <= MAX_JS_INT) {
+        ujson.Num(num.toDouble)
+      } else {
+        ujson.Obj("#bigint" -> ujson.Str(num.toString(10)))
+      }
+
+    case ValEx(TlaBool(b)) =>
+      ujson.Bool(b)
+
+    case ValEx(TlaStr(str)) =>
+      ujson.Str(str)
+
+    case ex @ OperEx(TlaFunOper.tuple, args @ _*) =>
+      ex.typeTag match {
+        case Typed(SeqT1(_)) =>
+          ujson.Arr(args.map(exToJson): _*)
+
+        case _ =>
+          ujson.Obj("#tup" -> ujson.Arr(args.map(exToJson): _*))
+      }
```

Comment:
```suggestion
      val arr = ujson.Arr(args.map(exToJson)
      ex.typeTag match {
        case Typed(SeqT1(_)) =>
          ujson.Arr(arr: _*)

        case _ =>
          ujson.Obj("#tup" -> arr: _*))
      }
```

Reply:
Good point. I did not notice that it was a list :)